### DISCUSSION TITLE
Why dont attach volume over API?

### DISCUSSION THREAD

**Discussion by Raiden1974** [longhorn/longhorn#3558] [Q&A]
## Question
Why dont attach volume over API? I creates volume in a project - all ok, but call ActionAttach cause timeout error.

## Environment

 - Longhorn version: 1.2.3
 - Kubernetes version: 1.22.5
 - Node config
   - Ubuntu 20
   -  CPU per node: 2
   - Memory per node: 4
   - Disk type SSD
   - 
## Additional context
My code:

`
func main() {
	client, err := c.NewRancherClient(&c.ClientOpts{
		Url:       "http://0.0.0.0/v1",
		Timeout:   30 * time.Second,
	})
	if err != nil {
		return
	}

	volume, err := client.Volume.Create(&c.Volume{
		AccessMode:         "rwo",
		Frontend:           "blockdev",
		DataLocality:       "disabled",
		Name:               "v2",
		NumberOfReplicas:   1,
		ReplicaAutoBalance: "ignored",
		Size:               "1Gi",
	})
	if err != nil {
		fmt.Println(err)
		return
	}

	nodeCollection, err := client.Node.List(nil)
	if err != nil {
		fmt.Println(err)
		return
	}

	time.Sleep(2 * time.Second)

	_, err = client.Volume.ActionAttach(volume, &c.AttachInput{
		DisableFrontend: false,
		HostId:          nodeCollection.Data[0].Id,
	})
	if err != nil {
		fmt.Println(err)
		return
	}
}
`

**Comment by innobead** [comment]
Because you are using our internal API, I would suggest checking longhorn-manager logs first to see what issues happened there. cc @PhanLe1010

**Comment by Raiden1974** [comment]
> Because you are using our internal API, I would suggest checking longhorn-manager logs first to see what issues happened there. cc @PhanLe1010

Post "http://10.42.0.9:9500/v1/volumes/v2?action=attach": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

**Comment by Raiden1974** [comment]
I try to debug you code and found what:
- When I create volume a action url is http://10.7.10.74/....
- When I do attach action url  http://10.42.0.9:9500/v1/volumes/v2?action=attach internal action map contains this address.

The 10.7.10.74 a host with k3s, 10.42.0.9 pod address for longhorn
For longhorn-frontend and API I make nginx ingress. If I use client API from other host, address of pod not reachable from client API source code, its true? 

<pre><font color="#4E9A06"><b>support@longhorn</b></font>:<font color="#3465A4"><b>~</b></font>$ kubectl get pods -n longhorn-system -o wide
NAME                                        READY   STATUS    RESTARTS   AGE     IP           NODE        NOMINATED NODE   READINESS GATES
longhorn-ui-9fdb94f9-k2fkc                  1/1     Running   0          3d7h    10.42.0.10   longhorn    &lt;none&gt;           &lt;none&gt;
longhorn-manager-jjgdk                      1/1     Running   0          3d7h    10.42.0.9    longhorn    &lt;none&gt;           &lt;none&gt;
</pre>

<pre><font color="#4E9A06"><b>support@longhorn</b></font>:<font color="#3465A4"><b>~</b></font>$ kubectl get pods -o wide
NAME                                READY   STATUS    RESTARTS   AGE     IP           NODE        NOMINATED NODE   READINESS GATES
longhorn-iscsi-installation-gt6lf   1/1     Running   0          3d6h    10.7.10.74   longhorn    &lt;none&gt;           &lt;none&gt;
longhorn-nfs-installation-hgrj5     1/1     Running   0          3d6h    10.7.10.74   longhorn    &lt;none&gt;           &lt;none&gt;
longhorn-nfs-installation-gsd5h     1/1     Running   0          3h41m   10.7.10.75   longhorn2   &lt;none&gt;           &lt;none&gt;
longhorn-iscsi-installation-rct2x   1/1     Running   0          3h41m   10.7.10.75   longhorn2   &lt;none&gt;           &lt;none&gt;
<font color="#4E9A06"><b>support@longhorn</b></font>:<font color="#3465A4"><b>~</b></font>$ kubectl get svc -o wide
kubectl get svc -o wide
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE    SELECTOR
kubernetes   ClusterIP   10.43.0.1    &lt;none&gt;        443/TCP   3d6h   &lt;none&gt;
<font color="#4E9A06"><b>support@longhorn</b></font>:<font color="#3465A4"><b>~</b></font>$ kubectl get svc -o wide
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE    SELECTOR
kubernetes   ClusterIP   10.43.0.1    &lt;none&gt;        443/TCP   3d6h   &lt;none&gt;
<font color="#4E9A06"><b>support@longhorn</b></font>:<font color="#3465A4"><b>~</b></font>$ kubectl -n longhorn-system get svc
NAME                       TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)     AGE
longhorn-backend           ClusterIP   10.43.248.27    &lt;none&gt;        9500/TCP    3d6h
longhorn-engine-manager    ClusterIP   None            &lt;none&gt;        &lt;none&gt;      3d6h
longhorn-replica-manager   ClusterIP   None            &lt;none&gt;        &lt;none&gt;      3d6h
longhorn-frontend          ClusterIP   10.43.220.49    &lt;none&gt;        80/TCP      3d6h
csi-attacher               ClusterIP   10.43.27.181    &lt;none&gt;        12345/TCP   3d6h
csi-provisioner            ClusterIP   10.43.182.166   &lt;none&gt;        12345/TCP   3d6h
csi-resizer                ClusterIP   10.43.75.88     &lt;none&gt;        12345/TCP   3d6h
csi-snapshotter            ClusterIP   10.43.34.103    &lt;none&gt;        12345/TCP   3d6h
</pre>

**Comment by PhanLe1010** [comment]
You are using the go-lang client which is not officially documented. Something need to be check:
* Your code should be running inside a pod in the cluster and the URL should be `http://longhorn-frontend.longhorn-system/v1`. If you are running this outside of the cluster you would need to expose the longhorn-frontend service 
* You need to specify `node.Name` instead if `node.Id` in the attach action. I.e., 
  ```
  _, err = client.Volume.ActionAttach(volume, &c.AttachInput{
	  DisableFrontend: false,
	  HostId:          nodeCollection.Data[0].Name,
  })
  ```
  see https://github.com/longhorn/longhorn-manager/blob/7ff0f32ae9050bb5ba1c5062973fc2bf48b2b374/client/generated_node.go#L22


Would you like to try the Python client instead. This client is documented at https://longhorn.io/docs/1.2.3/references/longhorn-client-python/. Our e2e testes are also using this client

**Comment by Raiden1974** [comment]
Node id has equal name value. Use python not possible  in our projects we only use Golang.

**Comment by PhanLe1010** [comment]
> Node id has equal name value

I don't think the node object return by `client.Node.List(nil)` has the field `Id`. If you take a look at ht node struct https://github.com/longhorn/longhorn-manager/blob/7ff0f32ae9050bb5ba1c5062973fc2bf48b2b374/client/generated_node.go#L7-L31

**Comment by Raiden1974** [comment]
I'm not talking about the structure, but in general about the code that I looked at. You write in the id the same as in the name field. The problem is not in this at all, but in that, even when creating a volume, the actions contain an address that is inaccessible from external hosts.

you can probably close this topic.

**Reply by PhanLe1010** [reply]
Okay. Thank you for the clarification about the Id and Name

If you are accessing from an external host, which URL are you using to create the NewRancherClient?

**Comment by Raiden1974** [comment]
The URL for create the NewRancherClient is http://[longhorn host ip]/v1

**Comment by PhanLe1010** [comment]
What is the `longhorn host ip`?

**Reply by Raiden1974** [reply]
It IP address of the host. Example - 10.7.10.71

**Comment by PhanLe1010** [comment]
Still looks like a problem with the URL for create the NewRancherClient to me. 

If you are still debugging, maybe you can try to do port forwarding the longhorn UI service `kubectl port-forward services/longhorn-frontend 8080:http -n longhorn-system` or create an ingress to longhorn-backend service
